### DISCUSSION TITLE
Is there a simple CDC example?

### DISCUSSION THREAD

**Discussion by greenscreenflicker** [hathach/tinyusb#2525] [General]
Hello,
i'm looking for a very basic CDC example, ideally only printing "Hello world" on the console.
Is there such examples? All i find are such complex...
Michael

**Comment by tyustli** [comment]
examples/device/cdc_dual_ports

**Reply by greenscreenflicker** [reply]
seen that, but what irritates me here, is that it's dual ports...

**Reply by tyustli** [reply]
Just change it.

![image](https://github.com/hathach/tinyusb/assets/43946994/388e85f4-49a4-45f0-af5e-d3a4ef63c72a)

**Reply by greenscreenflicker** [reply]
looks easy, will try that. thank you.

**Reply by greenscreenflicker** [reply]
I tryed. I have the same problem, that I don't see any usb cdc in the windows device manager.
Is there anything I need to do additonally, so that I see the USB Port?

My current task (and all the code I call) looks like this.

```
 void task2_task(void *pvParameters)
{
    tusb_init(); // initialize tinyusb stack

    int i=0;
    while(1) // the mainloop
    {
      tud_cdc_write_str("Hello, World!\r\n");
      if(i){
          GPIO_SetBits(GPIOB, GPIO_Pin_4);
          i=0;
      }else{
          GPIO_ResetBits(GPIOB, GPIO_Pin_4);
          i=1;
      }
      tud_task(); // device task
      vTaskDelay(100);
    }
}
```

**Reply by tyustli** [reply]
Make sure the enumeration is successful
![image](https://github.com/hathach/tinyusb/assets/43946994/88c75ca1-a112-4002-b2de-48ebf6c151dc)

**Reply by dragonlock2** [reply]
Looks like you're trying to use FreeRTOS. I believe `tud_task()` needs to be called in a continuous loop without the `vTaskDelay()` since that's what actually processes/queues USB packets. I'm assuming your device is failing to enumerate.

**Reply by greenscreenflicker** [reply]
I didn't get it to work. My code currently looks like that.
```
void task1_task(void *pvParameters)
{

    board_init();

    // init device stack on configured roothub port
    tud_init(BOARD_TUD_RHPORT);

    if (board_init_after_tusb) {
      board_init_after_tusb();
    }

    while (1) {
      tud_task(); // tinyusb device task
      tud_cdc_n_write_str(0,"Hello, World!\r\n");
      //vTaskDelay(10);
    }
}
```

I put all my code here, could somebody have a look and advise me how i could possible get it to work?

https://github.com/greenscreenflicker/ch32v20x_tinyusb_support/tree/main

I desperatly need to get usbcdc running for my project... It's a big roadblocker...

Edit: @dragonlock2: I'm failing to enumerate, you're right!

**Reply by dragonlock2** [reply]
Let's try to get to a place that we know will work. Try building and flashing one of the examples and see if it enumerates, something like the following.
```
cd tinyusb
python tools/get_deps.py ch32v20x
cd examples/device/hid_composite
make BOARD=nanoch32v203
```

**Reply by greenscreenflicker** [reply]
@dragonlock2 i use the moun river studio by wch, which is basically eclipse. http://www.mounriver.com/
I'm not sure how i should run there a console. 

What build environment do you use, so you can run the console?

**Reply by dragonlock2** [reply]
I use CMake on macOS personally. If you're on Windows it may be easier to build an example in WSL. You'll need the compiler on your path, examples recommend https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack

**Reply by greenscreenflicker** [reply]
Well, let me document my process:

1. [OK] I installed WSL2 on windows https://learn.microsoft.com/en-us/windows/wsl/install
2. [OK] Installed there ubuntu distribution, restarted system (windows asked for it), and then finished ubuntu installation
3. [ERRORLater]Followed the following tutorial to compile risc-v instruction set https://bernardnongpoh.github.io/posts/riscv
4.[ ERRORLater] It takes quite a time to compile - Grap a drink - it still takes -- i will update once finished.
5. [ERRORLater]It installed -64 but i need 32 bit for my mcu... bad...
6. [ERRORLater] followed this to install 32 bit version
7. [ERRORLater]https://stackoverflow.com/questions/74231514/how-to-install-riscv32-unknown-elf-gcc-on-debian-based-linuxes
8. [ERRORLater]Doesn't compile, going back to packages manager, as suggested by dragonlock2
9. [ERRORLater] Stopping that approach. Going von XPM
10. [XPM]fighting to install xpm, tutorial given on xpm install on doesn't work https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack
11. [XPM]finding instructions https://github.com/verseles/xpm: 'curl -sL xpm.link | sh' works
12. [XPM]Running `xpm init` returns "no packages found"
13. [XPM]Running `xpm install @xpack-dev-tools/riscv-none-elf-gcc@latest --verbose` returns "Package "@xpack-dev-tools/riscv-none-elf-gcc@latest" not found."
14. [XPM] Trying `xpm refresh`: Updates but doesn't find the package.
15. Any help how to get xpack-dev-tools installed?
16. Opened an issue in the installation: https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/issues/29

To be honest, I'm frustrated how hard it's to compile state-of-the-art software. It should be easier! But everyone thanks for helping so far.

**Reply by ilg-ul** [reply]
It is hard if you do not read the release pages to install the correct prerequisites:

- https://xpack.github.io/blog/2023/09/25/riscv-none-elf-gcc-v13-2-0-2-released/

and

- https://xpack.github.io

**Reply by greenscreenflicker** [reply]
I still have the problem @dragonlock2's code sample, that the compiler is not known. What do i wrong? Please support.
```
mmh@DESKTOP-5DHNM07:~/delte/tinyusb/examples/device/hid_composite$ xpm install --global @xpack-dev-tools/riscv-none-elf-gcc@latest --verbose --force
xPack manager - install package(s)

Processing @xpack-dev-tools/riscv-none-elf-gcc@13.2.0-2.1...
Changing permissions to read-write...
Removing existing package from '/home/mmh/.local/xPacks/@xpack-dev-tools/riscv-none-elf-gcc/13.2.0-2.1'...
Installing globally in '/home/mmh/.local/xPacks/@xpack-dev-tools/riscv-none-elf-gcc/13.2.0-2.1'...
Extracting 'xpack-riscv-none-elf-gcc-13.2.0-2-linux-x64.tar.gz'...
8059 files extracted in '13.2.0-2.1/.content'
Changing permissions to read-only...

'xpm install' completed in 81.737 sec.
mmh@DESKTOP-5DHNM07:~/delte/tinyusb/examples/device/hid_composite$ make BOARD=nanoch32v203
make: riscv-none-elf-gcc: No such file or directory
expr: syntax error: unexpected argument ‘12’
AS startup_ch32v20x_D6_asm.o
make: riscv-none-elf-gcc: No such file or directory
make: *** [/home/mmh/delte/tinyusb/examples/build_system/make/toolchain/arm_gcc_rules.mk:62: _build/nanoch32v203/obj/hw/mcu/wch/ch32v20x/EVT/EXAM/SRC/Startup/startup_ch32v20x_D6_asm.o] Error 127
```

**Reply by dragonlock2** [reply]
`xpm` doesn't add the installed GCC to your `PATH` by default. It's probably somewhere inside `/home/mmh/.local/xPacks/@xpack-dev-tools/riscv-none-elf-gcc/13.2.0-2.1`, modify your `PATH` accordingly. If you can run `riscv-none-elf-gcc` from your shell, you'll know you did it correctly.

**Reply by greenscreenflicker** [reply]
Well, there's no gcc. 

> mmh@DESKTOP-5DHNM07:~/.local/xPacks/@xpack-dev-tools/riscv-none-elf-gcc/13.2.0-2.1$ ls
> 
> CHANGELOG.md  LICENSE  README.md  package.json

EDIT: Where to find the extracted binary?

**Reply by dragonlock2** [reply]
`ls -a` since the logs says it extracted stuff into `.content` (hidden folder).

**Reply by greenscreenflicker** [reply]
> mmh@DESKTOP-5DHNM07:~/.local/xPacks/@xpack-dev-tools/riscv-none-elf-gcc/13.2.0-2.1$ ls -a
> .  ..  .content  CHANGELOG.md  LICENSE  README.md  package.json

EDIT: Found it in .content

**Reply by greenscreenflicker** [reply]
Okay, i added the path and the compile was successfull. Thank you.
How do I get the bin to windows, so i can flash it?

EDIT: My path command was: 

> export PATH="/home/mmh/
.local/xPacks/@xpack-dev-tools/riscv-none-elf-gcc/13.2.0-2.1/.content/bin/:$PATH"

**Reply by greenscreenflicker** [reply]
@dragonlock2 I tried the compiled code. When compiling it looks to me that I have the wrong mcu defined. I have only one with 32kB flash, not 64kbyte. How do I change that?

Note: i found the https://github.com/dragonlock2/tinyusb/blob/master/hw/bsp/ch32v20x/boards/nanoch32v203/board.mk
but i don't know what to change...

**Reply by dragonlock2** [reply]
Change the CH32V20x_D6 parts to your variant. If the flash usage of the app is below 32KB, the same binary should just work for both variants I would just try that first.

edit: nvm the 32KB has less sram, do modify the makefile bc the stack will be in the wrong place

**Reply by greenscreenflicker** [reply]
Well, first i tried flashing directly, usb is not enumereated. Strangly when compiling it shows me 20kb ram usage, thats why my guess is the mcu is wring defined.

I use the CH32V203G6U6. But CH32V20x_G6 doesnt work.

Mcu link
https://www.lcsc.com/product-detail/Microcontroller-Units-MCUs-MPUs-SOCs_WCH-Jiangsu-Qin-Heng-CH32V203G6U6_C5142280.html

**Reply by dragonlock2** [reply]
Look into the ch32v20x SDK which is a submodule. D6 is correct it seems, but the linker script needs to be modified bc it defaults to 64KB with 32KB commented out. Linker script location is mentioned in family.mk

**Reply by greenscreenflicker** [reply]
Okay, I compiled it with D6. Modified linker script to 32kB flash, 10kB Ram, changed oszillator to internal (do not have an external crystal connected. Flashed it with the generated bin file. Doesn't get enumerated.
As it's not in eclipse/moun river studio i cannot debug it well.

What else could i do?

Desperately tried to flash the hex file. Same result.

**Reply by dragonlock2** [reply]
I realized the CH32V203G6U6 doesn't have the USB OTG/FS peripheral (which my driver is written for) but instead only has the USBD peripheral. I don't have plans to work on support for that at the moment.

However, briefly looking at the register maps of the two I'm pretty sure USBD is a copy of the STM32 USB peripheral which means it's possible to get that driver [dcd_stm32_fsdev.c](https://github.com/hathach/tinyusb/blob/master/src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c) working with the right set of headers and macros.

If you want to check if your hardware is working, you can try getting into the bootloader which should enumerate on both USBD and USBOTG_FS ports, although the lack of a crystal might mess things up.

**Reply by greenscreenflicker** [reply]
sadly, i only have that mcu here now. So i need to look for another solution.

**Reply by dragonlock2** [reply]
As mentioned https://github.com/hathach/tinyusb/pull/2362#issuecomment-2024320112, here's a project that can be configured for either USBFS or USBD of the CH32V20x [dragonlock2/miscboards/wch/mouse_jiggler](https://github.com/dragonlock2/miscboards/tree/main/wch/mouse_jiggler). As I expected, the dcd_stm32_fsdev.c driver works for CH32V20x too because they literally copied the register map. The project should fit within your 32KB part, except you'll change to the HSI.

I'll leave the porting of my example project to your build system as a learning exercise for you, good luck.

**Reply by ddB0515** [reply]
@dragonlock2 thanks for sample, I have compiled it but I noticed it's for v203 with 64KB flash and 20KB RAM (like CH32V203C8U6)

Memory region         Used Size  Region Size  %age Used
             ROM:       23348 B        64 KB     35.63%
             RAM:        7044 B        20 KB     34.39%

I'm new to tinyusb and not sure that MCU can be defined for different size linker script to support 32K/10K (CH32V203G6U6)
As from builded sample it will fit (still room for ~6KB FLASH and ~3KB RAM

**Reply by greenscreenflicker** [reply]
@ddB0515 look for the Flash.ld script above, and comment in the right sizes.

**Reply by ddB0515** [reply]
I know that :) just asking if there is any alternative for declaring different MCU to change this automatically like in WCH SDK

**Reply by greenscreenflicker** [reply]
@ddB0515 which WCH-SDK are you using?

**Reply by ddB0515** [reply]
In my other project I'm using `noneos-sdk` with VSCode and PlatformIO but came for tinyusb ease of use and as I can see it's using FreeRTOS so need to get used to, trying now to change this example to work on CH32V203G6U6 and will try on EVT board from WCH and post results once I managed to get it running

**Reply by ddB0515** [reply]
so I have managed to get my EVT with CH32V203C8T6 (64Kb/20Kb) to flash with compiled sample 
I didn't get further than
`waiting for usb mount...`

Also checked 
tud_mounted(): false
tud_suspended(): false
tud_hid_ready(): false

so for a reason I'm not getting after this check 
`while (!tud_mounted() || tud_suspended() || !tud_hid_ready());`

One thing that I still trying to figure it out what USB pins are used? 
There is 2 USB ports (first is USB C connector to connect to PC)
PA12/USB1DP
PA11/USB1DM
and 2nd connector on this EVT (for HostDevice) 
PB7/USB2DP
PB6/USB2DM

![image](https://github.com/hathach/tinyusb/assets/20762436/c588a9b0-1f8a-4fa9-9a11-78fd087a2212)


As I'm new to tinyusb not sure where this is set? Or I'm missing something :thinking:

**Reply by dragonlock2** [reply]
I use the nanoCH32V203 personally so not sure about this board's connectors. PA11/12 should be USBD. PB6/PB7 should be USBFS

**Reply by ddB0515** [reply]
Checked schematic and same pins are on both boards and I'm using PA11/12 for USBD. Does example works on your board normally? No issues?

**Reply by dragonlock2** [reply]
Looking at the schematic [CH32V203C8T6-R0.pdf](https://github.com/openwch/ch32v20x/blob/main/SCHPCB/CH32V203C-R0/SCH_PCB/CH32V203C8T6-R0.pdf), PA11/PA12 are actually on the P6 connector, labeled USB2(Host) in the image. You'll need a USB-A to USB-A cable or something similar.

On the other hand, the USB-C connector has R22/R23 DNP by default so you can't even test the USBFS (using the USB-C connector) without soldering them on or using a USB-A to USB-C cable.

**Reply by ddB0515** [reply]
OMG you are right, so all this time it was working just used wrong USB port :man_facepalming: 
Connected using USB-A to USB-A it's working...

[31890.237801] input: miscboards mouse_jiggler as /devices/pci0000:00/0000:00:02.1/0000:04:00.0/0000:05:0c.0/0000:15:00.0/usb3/3-7/3-7.4/3-7.4.1/3-7.4.1:1.0/0003:CAFE:0069.003E/input/input56
[31890.300626] hid-generic 0003:CAFE:0069.003E: input,hidraw8: USB HID v1.11 Mouse [miscboards mouse_jiggler] on usb-0000:15:00.0-7.4.1/input0

And mouse is recognized by Hidviz
![image](https://github.com/hathach/tinyusb/assets/20762436/11d8327f-cf2d-4fef-b6fd-d81d5d7c377a)

Next task is to try CH32V203G6U6 and this one should be easier as is only USBD without USBFS and having only 1 USB :)
Will post here update/progress once I manage to change linker/size and try...
Thanks a lot for support :beers:

**Reply by ddB0515** [reply]
WOW 

This was so easy for CH32V203G6U6 (changed Linker script to 32K/10K)
Deleted `build` folder, clean build and flash

[35443.850672] usb 3-7.2: new full-speed USB device number 39 using xhci_hcd
[35444.053768] usb 3-7.2: New USB device found, idVendor=cafe, idProduct=0069, bcdDevice= 0.00
[35444.053774] usb 3-7.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[35444.053777] usb 3-7.2: Product: mouse_jiggler
[35444.053778] usb 3-7.2: Manufacturer: miscboards
[35444.053780] usb 3-7.2: SerialNumber: 69420
[35444.068302] input: miscboards mouse_jiggler as /devices/pci0000:00/0000:00:02.1/0000:04:00.0/0000:05:0c.0/0000:15:00.0/usb3/3-7/3-7.2/3-7.2:1.0/0003:CAFE:0069.0048/input/input64
[35444.068397] hid-generic 0003:CAFE:0069.0048: input,hidraw8: USB HID v1.11 Mouse [miscboards mouse_jiggler] on usb-0000:15:00.0-7.2/input0

Also nice show of usage and plenty of room :)
Memory region        Size        Region Size        %age Used
ROM:        23684 B        32 KB     72.28%
RAM:        7044 B            10 KB     68.79%

Such nice to have this working on both WCH chips

@dragonlock2  you have a :beer:  or :beers:

**Comment by HiFiPhile** [comment]
https://googlethatforyou.com?q=how%20do%20i%20transfer%20files%20from%20wsl%20to%20windows

**Comment by ilg-ul** [comment]
**xpm** does not add anything to the PATH, since, by design, the xPack framework allows to use different versions of a toolchain in different projects.

Instead xpm adds links to the `xpacks/.bin` folder, which you should add to the local project PATH.

You can automate this by using xPack _actions_, which are invoked via `xpm run <name>`; in this case `xpacks/.bin` is automatically added to the PATH.

There is also a VS Code extension allowing to conveniently run these actions with a click of the mouse.

If you save `package.json` to your repository, if you clone it on another machine, you can install _exactly the same version_ of the toolchain via `xpm install`. This also works on Windows directly, you don't have to use WSL. Actually it works the same on all platforms, Windows, macOS and Linux.
